lib/BasicAnalyses/CSVAliasAnalysis.cpp

```diff
@@ -0,0 +1,259 @@
+  ConstantInt *Offset = nullptr;
+  Value *Ptr = nullptr;
+
+  if (skipCasts(Ptr = I->getOperand(0)) == SP)
+    Offset = dyn_cast<ConstantInt>(I->getOperand(1));
+  else if (skipCasts(Ptr = I->getOperand(1)) == SP)
+    Offset = dyn_cast<ConstantInt>(I->getOperand(0));
+
+  if (Ptr && Offset) {
+    IRBuilder<> Builder(I);
+    auto *ITP = createOpaqueIntToPtr(Builder, Ptr);
+    auto *GEP = Builder.CreateGEP(Builder.getInt8Ty(), ITP, Offset);
+    auto *PTI = Builder.CreatePtrToInt(GEP, I->getType());
+    I->replaceAllUsesWith(PTI);
+  }
+}
+
+template<bool SegregateStackAccesses>
+using CSVAAP = CSVAliasAnalysisPass<SegregateStackAccesses>;
+
+template<bool SegregateStackAccesses>
+void CSVAAP<SegregateStackAccesses>::initializeAliasInfo(Module &M) {
+  LLVMContext &Context = M.getContext();
+  QuickMetadata QMD(Context);
+  MDBuilder MDB(Context);
+
+  AliasDomain = MDB.createAliasScopeDomain("revngAliasDomain");
+
+  std::set<GlobalVariable *> CSVs;
+  NamedMDNode *NamedMD = M.getOrInsertNamedMetadata("revng.csv");
+  auto *Tuple = cast<MDTuple>(NamedMD->getOperand(0));
+  for (const MDOperand &Operand : Tuple->operands()) {
+    auto *CSV = cast<GlobalVariable>(QMD.extract<Constant *>(Operand.get()));
+    CSVs.insert(CSV);
+  }
+
+  const auto *PCH = GCBI->programCounterHandler();
+  for (GlobalVariable *PCCSV : PCH->pcCSVs())
+    CSVs.insert(PCCSV);
+
+  // Build alias scopes
+  for (const GlobalVariable *CSV : CSVs) {
+    CSVAliasInfo &AliasInfo = CSVAliasInfoMap[CSV];
+
+    std::string Name = CSV->getName();
+    MDNode *CSVScope = MDB.createAliasScope(Name, AliasDomain);
+    AliasInfo.AliasScope = CSVScope;
+    AllCSVScopes.push_back(CSVScope);
+    MDNode *CSVAliasSet = MDNode::get(Context,
+                                      ArrayRef<Metadata *>({ CSVScope }));
+    AliasInfo.AliasSet = CSVAliasSet;
+  }
+
+  // Build noalias sets
+  for (const GlobalVariable *CSV : CSVs) {
+    CSVAliasInfo &AliasInfo = CSVAliasInfoMap[CSV];
+    std::vector<Metadata *> OtherCSVScopes;
+    for (const auto &Q : CSVAliasInfoMap)
+      if (Q.first != CSV)
+        OtherCSVScopes.push_back(Q.second.AliasScope);
+
+    MDNode *CSVNoAliasSet = MDNode::get(Context, OtherCSVScopes);
+    AliasInfo.NoAliasSet = CSVNoAliasSet;
+  }
+}
+
+template<bool SegregateStackAccesses>
+void CSVAAP<SegregateStackAccesses>::decorateMemoryAccesses(Instruction &I) {
+  Value *Ptr = nullptr;
+
+  if (auto *L = dyn_cast<LoadInst>(&I))
+    Ptr = L->getPointerOperand();
+  else if (auto *S = dyn_cast<StoreInst>(&I))
+    Ptr = S->getPointerOperand();
+  else
+    return;
+
+  // Check if the pointer is a CSV
+  if (auto *GV = dyn_cast<GlobalVariable>(Ptr)) {
+    auto It = CSVAliasInfoMap.find(GV);
+    if (It != CSVAliasInfoMap.end()) {
+      // Set alias.scope and noalias metadata
+      I.setMetadata(LLVMContext::MD_alias_scope,
+                    MDNode::concatenate(I.getMetadata(LLVMContext::MD_noalias),
```

Comment:
```suggestion
                    MDNode::concatenate(I.getMetadata(LLVMContext::MD_alias_scope),
```

Again? Is this the latest version of the code?